### PR TITLE
Add support for nullable in column

### PR DESCRIPTION
**Pull request by Toflar**
|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | -

#### Summary

I just can't wrap my head around `'notnull' => false`. I find `'nullable' => true` much more intuitive. Probably because that's how we use it in ORM. So I figured I just create this PR to see if this would improve DX a little and help others :D

### CODE REVIEW

**Comment by derrabus** [comment]
> I just can't wrap my head around `'notnull' => false`.

When working with SQL databases, `NOT NULL` is pretty much a common concept. 😉 

> I find `'nullable' => true` much more intuitive.

I understand that. Yet your increases the complexity of your codebase and I'm not convinced it's worth it.

* What happens if both `notnull` and `nullable` are set, maybe even to conflicting values?
* How does the new settings affect diffing tables? Will columns still be considered equal if one is declared using `notnull` and the other one using `nullable`?

Note that we do not only need to answer those question, but we need tests that cover those scenarios. You've only tested the happy path so far.

**Comment by Toflar** [comment]
> When working with SQL databases, `NOT NULL` is pretty much a common concept. 😉

Yes of course. I don't know what it is. It just feels strange 😄 Maybe it's because in a regular database, if I want to have a nullable column, I have to **add** `NULL` to the `ALTER` statement? And "adding" something with `'notnull' => false` feels unintuitive? I know, I also have to add `NOT NULL` - I'm just trying to find out why one feels better than the other 😉 

> What happens if both notnull and nullable are set, maybe even to conflicting values?

Well, the later just wins. The same as if you set the same key multiple times - the latter overrides the former.

> How does the new settings affect diffing tables? Will columns still be considered equal if one is declared using notnull and the other one using nullable?

No changes. The internal array is the same as now. There is no new `nullable` option. It remains `notnull` everywhere. It's only used in `Column::setOptions()` and thus I cannot think of any possible side-effects 🤔

**Comment by derrabus** [comment]
> The same as if you set the same key multiple times - the latter overrides the former.

Well, I cannot have the same key in one array twice, can I.

> No changes. The internal array is the same as now.

I see that. But again:

> Note that we do not only need to answer those question, but we need tests that cover those scenarios.

**Comment by Toflar** [comment]
> Well, I cannot have the same key in one array twice, can I.

Yes, you can? https://3v4l.org/2hsbN#v8.2.3
But we can also throw an exception, if you prefer?

> Note that we do not only need to answer those question, but we need tests that cover those scenarios.

Sure! I've added tests in `ComparatorTest` - any other specific scenario you're thinking about?

**Comment by derrabus** [comment]
> > Well, I cannot have the same key in one array twice, can I.
> 
> Yes, you can? https://3v4l.org/2hsbN#v8.2.3

Not sure if I get a different output than you, but the resulting array does not have duplicate keys, no matter what you write into your array constructor expression.

**Comment by Toflar** [comment]
I don't quite get your point, maybe we talk at cross purposes 😄 Anyway, I've added an exception because I agree it's better anyway :)

**Comment by derrabus** [comment]
> I don't quite get your point

I think you are confusing an array constructor with an array. An array constructor is an expression that you can use to construct an array. The language allows you to specify the same key twice here. But the actual array that is constructed won't have a duplicate key.

If I'm holding an array in my hand, the runtime guarantees that each key is unique.

File: src/Schema/Column.php [hunk 1]
```diff
@@ -82,7 +85,20 @@ public function __construct($name, Type $type, array $options = [])
...
+        if (
+            array_key_exists('notnull', $options)
+            && array_key_exists('nullable', $options)
+            && $options['notnull'] === $options['nullable']
+        ) {
+            throw ContradictingColumnOption::new('notnull', 'nullable');
+        }
+
         foreach ($options as $name => $value) {
+            if ($name === 'nullable' && is_bool($value)) {
```
Comment (derrabus) [inline src/Schema/Column.php]: And if it's not a bool? I get a `UnknownColumnOption` exception? 🤨

**Comment by derrabus** [comment]
Sorry, the more I think about this change, the less do I want to have that code in the DBAL codebase. We're generating an unnecessary edge case with no real gain.

I'm closing this now. But I thank you for your attempt at improving DBAL.

**Pull request closed without merging**